src/Management/src/Endpoint/Actuators/HttpExchanges/HttpExchangesDiagnosticObserver.cs

```diff
@@ -0,0 +1,202 @@
+            requestUri += context.Request.QueryString.Value;
+        }
+
+        string? remoteAddress = null;
+
+        if (options.AddRemoteAddress)
+        {
+            remoteAddress = GetRemoteAddress(context);
+        }
+
+        Dictionary<string, IList<string?>> requestHeaders = [];
+
+        if (options.AddRequestHeaders)
+        {
+            requestHeaders = GetHeaders(context.Request.Headers, options.RequestHeaders);
+        }
+
+        var traceRequest = new TraceRequest(context.Request.Method, requestUri, requestHeaders, remoteAddress);
+
+        Dictionary<string, IList<string?>> responseHeaders = [];
+
+        if (options.AddResponseHeaders)
+        {
+            responseHeaders = GetHeaders(context.Response.Headers, options.ResponseHeaders);
+        }
+
+        var traceResponse = new TraceResponse(context.Response.StatusCode, responseHeaders);
+
+        string? userName = null;
+
+        if (options.AddUserPrincipal)
+        {
+            userName = GetUserPrincipal(context);
+        }
+
+        TracePrincipal? principal = userName == null ? null : new TracePrincipal(userName);
+
+        string? sessionId = null;
+
+        if (options.AddSessionId)
+        {
+            sessionId = GetSessionId(context);
+        }
+
+        TraceSession? session = sessionId == null ? null : new TraceSession(sessionId);
+
+        long timestamp = GetJavaTime(DateTime.UtcNow.Ticks);
+
+        return new HttpExchange(traceRequest, traceResponse, timestamp, principal, session, duration.TotalMilliseconds);
+    }
+
+    internal static HttpContext? GetHttpContextPropertyValue(object instance)
+    {
+        return GetPropertyOrDefault<HttpContext>(instance, "HttpContext");
+    }
+
+    internal static Dictionary<string, IList<string?>> GetHeaders(IHeaderDictionary headers, IList<string> allowedHeaders)
+    {
+        var result = new Dictionary<string, IList<string?>>();
+
+        foreach (KeyValuePair<string, StringValues> pair in headers)
+        {
+#pragma warning disable S4040 // Strings should be normalized to uppercase
+            result.Add(pair.Key.ToLowerInvariant(), HeaderShouldBeRedacted(pair.Key, allowedHeaders) ? new StringValues(Redacted) : pair.Value);
+#pragma warning restore S4040 // Strings should be normalized to uppercase
+        }
+
+        return result;
+    }
+
+    private static bool HeaderShouldBeRedacted(string currentHeader, IList<string> allowedHeaders)
+    {
+        return !allowedHeaders.Contains(currentHeader) || (allowedHeaders.Count == 1 && string.IsNullOrEmpty(allowedHeaders[0]));
+    }
+
+    private void RecordHttpTrace(Activity current, HttpContext context)
+    {
+        ArgumentNullException.ThrowIfNull(current);
+        ArgumentNullException.ThrowIfNull(context);
+
+        HttpExchange trace = GetHttpExchange(context, current.Duration);
+        Queue.Enqueue(trace);
+
+        if (Queue.Count > _optionsMonitor.CurrentValue.Capacity && !Queue.TryDequeue(out _))
```

Comment:
Why would dequeue fail when an entry was just added?

Reply:
I'm not sure, I didn't change this code

Reply:
This happens when `options.Capacity` is set to a negative value. The configurer should prevent that from happening, and then this check can be removed.

Reply:
This line is how the queue is limited to `options.Capacity` - I think we need everything here except the log statement... but is it worth removing? It _shouldn't_ ever execute, but it might be interesting if it does

Reply:
This is unreachable code now, so I don't see why we should preserve that. It's just raising questions on why it exists when examined. Suggested change:
```c#
if (Queue.Count > _optionsMonitor.CurrentValue.Capacity)
{
    Queue.TryDequeue(out _);
}
```

Multiple threads may be running this code concurrently, so there's a potential window where the queue gets slightly larger than the maximum, for a very short time, but that shouldn't be a problem.